node/libs/crypto/src/bn254/mod.rs

```diff
@@ -0,0 +1,213 @@
+        self.0.serialize_compressed(&mut buf).unwrap();
+        buf
+    }
+}
+
+/// Type safety wrapper around G2.
+#[derive(Clone, PartialEq, Eq)]
+pub struct PublicKey(pub G2);
+
+impl ByteFmt for PublicKey {
+    fn decode(bytes: &[u8]) -> anyhow::Result<Self> {
+        G2::deserialize_compressed(bytes)
+            .map(Self)
+            .map_err(|e| anyhow!("failed to decode public key: {e:?}"))
+    }
+
+    fn encode(&self) -> Vec<u8> {
+        let mut buf = Vec::new();
+        self.0.serialize_compressed(&mut buf).unwrap();
+        buf
+    }
+}
+
+impl std::hash::Hash for PublicKey {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.0.hash(state);
+    }
+}
+
+impl PartialOrd for PublicKey {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for PublicKey {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        ByteFmt::encode(self).cmp(&ByteFmt::encode(other))
+    }
+}
+
+/// Type safety wrapper around a G1 value.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Signature(pub G1);
+
+impl Signature {
+    /// Verifies a signature against the provided public key
+    #[inline(never)]
+    pub fn verify(&self, msg: &[u8], pk: &PublicKey) -> Result<(), Error> {
+        let hash_point = hash::hash_to_g1(msg);
+
+        // First pair: e(H(m): G1, pk: G2)
+        let a = Bn254::pairing(hash_point, pk.0);
+        // Second pair: e(sig: G1, generator: G2)
+        let b = Bn254::pairing(self.0, G2::generator());
+
+        if a == b {
+            Ok(())
+        } else {
+            Err(Error::SignatureVerificationFailure)
+        }
+    }
+}
+
+impl ByteFmt for Signature {
+    fn decode(bytes: &[u8]) -> anyhow::Result<Self> {
+        G1::deserialize_compressed(bytes)
+            .map(Self)
+            .map_err(|e| anyhow!("failed to decode signature: {e:?}"))
+    }
+    fn encode(&self) -> Vec<u8> {
+        let mut buf = Vec::new();
+        self.0.serialize_compressed(&mut buf).unwrap();
+        buf
+    }
+}
+
+impl PartialOrd for Signature {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Signature {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        ByteFmt::encode(self).cmp(&ByteFmt::encode(other))
+    }
+}
+/// Type safety wrapper around [Signature] indicating that it is an aggregated signature.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct AggregateSignature(pub G1);
+
+impl AggregateSignature {
+    /// Generates an aggregate signature from a list of signatures.
+    pub fn aggregate<'a>(sigs: impl IntoIterator<Item = &'a Signature>) -> Self {
+        let mut agg = G1::zero();
+        for sig in sigs {
+            agg += sig.0
+        }
+
+        AggregateSignature(agg)
+    }
+
+    #[inline(never)]
+    fn verify_raw(&self, msgs_and_pks: &[(&[u8], &PublicKey)]) -> Result<(), Error> {
```

Comment:
Random question just to clarify: Do I understand correctly that there are proofs of knowledge are available for the aggregated `PublicKey`s to prevent rogue key attacks?

Reply:
I understood from @brunoffranca that this attack is not applicable, but i'm not sure whether proofs of knowledge are used. 

@brunoffranca

Reply:
Before a block is signed by anyone, the set of public keys of valid signers is already fixed.

Reply:
Yes, there will be PoPs. We will have a validator registry on L1 and you'll be required to provide a PoP to have your key included.